### PR TITLE
[DoctrineBridge][PropertyInfo] Added support for Doctrine Embeddables

### PR DESCRIPTION
**Pull request by vudaltsov**
| Q             | A
| ------------- | ---
| Branch?       | 2.8 and higher
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

Note that [Embeddables appeared only in doctrine/orm 2.5](http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/changelog/migration_2_5.html). I added class_exists checks for that.

### CODE REVIEW

File: src/Symfony/Bridge/Doctrine/Tests/PropertyInfo/DoctrineExtractorTest.php [hunk 1]
```diff
@@ -70,6 +70,27 @@ public function testExtract($property, array $type = null)
         $this->assertEquals($type, $this->extractor->getTypes('Symfony\Bridge\Doctrine\Tests\PropertyInfo\Fixtures\DoctrineDummy', $property, array()));
     }
 
+    public function testExtractEmbeddable()
+    {
+        if (!class_exists('Doctrine\ORM\Mapping\Embedded')) {
+            return;
+        }
+
+        $expectedTypes = [new Type(
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Doctrine/Tests/PropertyInfo/DoctrineExtractorTest.php]: please use array() instead of []

Comment (vudaltsov) [inline src/Symfony/Bridge/Doctrine/Tests/PropertyInfo/DoctrineExtractorTest.php]: thank you, changed PHP version to 5.3 in PhpStorm for this project and fixed declaration

**Comment by vudaltsov** [comment]
Suddenly I faced another problem with embeddables.

When an entity has properties with embeddables, `Doctrine\Common\Persistence\Mapping\ClassMetadata::getFieldNames()` returns property paths for them. Example:

```php
/**
 * @ORM\Entity
 */
class DoctrineWithEmbedded
{
    /**
     * @Id
     * @Column(type="smallint")
     */
    public $id;

    /**
     * @Embedded(class="DoctrineEmbeddable")
     */
    protected $embedded;
}

/**
 * @ORM\Embeddable
 */
class DoctrineEmbeddable
{
    /**
     * @Column(type="string")
     */
    protected $field;
}

$container->get('doctrine.orm.default_entity_manager.metadata_factory')
    ->getMetadataFor('DoctrineWithEmbedded')
    ->getFieldNames();

// returns
[
    'id',
    'embedded.field',
];
```

Fixed this in the next commit.

File: src/Symfony/Bridge/Doctrine/Tests/PropertyInfo/DoctrineExtractorTest.php [hunk 2]
```diff
@@ -62,6 +62,21 @@ public function testGetProperties()
         );
     }
 
+    public function testGetEmbeddableProperties()
+    {
+        if (!class_exists('Doctrine\ORM\Mapping\Embedded')) {
+            return;
```
Comment (stloyd) [inline src/Symfony/Bridge/Doctrine/Tests/PropertyInfo/DoctrineExtractorTest.php]: You should put `$this->markTestSkipped()` with message that embedded is not available.

File: src/Symfony/Bridge/Doctrine/Tests/PropertyInfo/DoctrineExtractorTest.php [hunk 3]
```diff
@@ -70,6 +85,27 @@ public function testExtract($property, array $type = null)
         $this->assertEquals($type, $this->extractor->getTypes('Symfony\Bridge\Doctrine\Tests\PropertyInfo\Fixtures\DoctrineDummy', $property, array()));
     }
 
+    public function testExtractEmbeddable()
+    {
+        if (!class_exists('Doctrine\ORM\Mapping\Embedded')) {
+            return;
```
Comment (stloyd) [inline src/Symfony/Bridge/Doctrine/Tests/PropertyInfo/DoctrineExtractorTest.php]: Same here.

**Review by stloyd — CHANGES_REQUESTED** [review]

File: src/Symfony/Bridge/Doctrine/Tests/PropertyInfo/DoctrineExtractorTest.php (hunk 2 above)
Comment (vudaltsov) [inline src/Symfony/Bridge/Doctrine/Tests/PropertyInfo/DoctrineExtractorTest.php]: Thank you, fixed this!

**Review by stloyd — APPROVED** [review]

**Comment by vudaltsov** [comment]
@nicolas-grekas, anything else needed here? could it be merged?

**Comment by dunglas** [comment]
This should be merged in 3.4 (it's a new feature, not a bug fix).

File: src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php [hunk 4]
```diff
@@ -50,7 +50,19 @@ public function getProperties($class, array $context = array())
             return;
         }
 
-        return array_merge($metadata->getFieldNames(), $metadata->getAssociationNames());
+        $properties = array_merge($metadata->getFieldNames(), $metadata->getAssociationNames());
+
+        if (class_exists('Doctrine\ORM\Mapping\Embedded')) {
+            if ($metadata instanceof ClassMetadataInfo && count($metadata->embeddedClasses) > 0) {
```
Comment (dunglas) [inline src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php]: `&& $metadata->embeddedClasses` will be faster.

File: src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php [hunk 5]
```diff
@@ -50,7 +50,19 @@ public function getProperties($class, array $context = array())
             return;
         }
 
-        return array_merge($metadata->getFieldNames(), $metadata->getAssociationNames());
+        $properties = array_merge($metadata->getFieldNames(), $metadata->getAssociationNames());
+
+        if (class_exists('Doctrine\ORM\Mapping\Embedded')) {
```
Comment (dunglas) [inline src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php]: As we'll merge this PR in 3.4, you can import this class and do `class_exists(Embedded::class)`.

File: src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php [hunk 6]
```diff
@@ -105,6 +117,16 @@ public function getTypes($class, $property, array $context = array())
             ));
         }
 
+        if (class_exists('Doctrine\ORM\Mapping\Embedded')) {
```
Comment (dunglas) [inline src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php]: same here, same in tests.

**Review by dunglas — CHANGES_REQUESTED** [review]
Nice improvement! I just left some minor comments.

**Comment by vudaltsov** [comment]
@dunglas, the problem is that now when you do
```php
$container
    ->get('doctrine.orm.default_entity_manager.property_info_extractor')
    ->getProperties(DoctrineWithEmbedded::class); // the class from my example above
```
you receive 
```php
[
    'id',
    'embedded.field', // not embedded
];
```
which is not what developer might expect.

So it appears that currently 2.8 doesn't fully support ORM, which might lead to unexpected behaviour. Isn't it a bug?

**Comment by dunglas** [comment]
Ok to consider it a bug fix then. 👍 

ping @symfony/deciders

**Review by dunglas — APPROVED** [review]

**Comment by vudaltsov** [comment]
@symfony/deciders, anything else needed here?

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php [hunk 7]
```diff
@@ -50,7 +50,17 @@ public function getProperties($class, array $context = array())
             return;
         }
 
-        return array_merge($metadata->getFieldNames(), $metadata->getAssociationNames());
+        $properties = array_merge($metadata->getFieldNames(), $metadata->getAssociationNames());
+
+        if (class_exists('Doctrine\ORM\Mapping\Embedded') && $metadata instanceof ClassMetadataInfo && $metadata->embeddedClasses) {
```
Comment (dunglas) [inline src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php]: Can you call class_exists after the other tests (for perf, function calls cost more than other checks)?

Comment (vudaltsov) [inline src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php]: @dunglas, when class `Doctrine\ORM\Mapping\Embedded` doesn't exist, `$metadata->embeddedClasses` is not defined. I can only move `instanceof` check to the first place.

Comment (dunglas) [inline src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php]: Ok can you do that? Then I'll merge :)

Comment (vudaltsov) [inline src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php]: @dunglas, done!

**Comment by dunglas** [comment]
Thank you @vudaltsov.

**Pull request closed without merging**